eclipse/src/saros/session/resources/validation/ResourceChangeValidator.java

```diff
@@ -0,0 +1,97 @@
+package saros.session.resources.validation;
+
+import org.apache.log4j.Logger;
+import org.eclipse.core.resources.IResourceDelta;
+import org.eclipse.core.resources.mapping.ModelProvider;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import saros.Saros;
+import saros.session.ISarosSession;
+import saros.session.User.Permission;
+
+/**
+ * This model provider is responsible for warning a session participant when trying to modify the
+ * resources of a shared project or the project itself in an unsupported way.
+ *
+ * <p>Ignoring the warnings will lead to inconsistencies. Note that we can't prevent these actions,
+ * just detect them and report them.
+ *
+ * <p>Currently detected:
+ *
+ * <ul>
+ *   <li>File and folder modifications as a user with {@link Permission#READONLY_ACCESS}.
+ *   <li>Deletion of a shared project.
+ *   <li>Renaming and moving of a shared project.
+ * </ul>
+ */
+/*
+ * FIXME there are several deltas that report changes which do not affect the
+ * file system. As this class did not work until Eclipse 4.x ??? nobody ever
+ * realized this
+ */
+public class ResourceChangeValidator extends ModelProvider {
+  private static final Logger log = Logger.getLogger(ResourceChangeValidator.class);
+
+  /** Error code for internal use, but we don't need it. */
+  private static final int ERROR_CODE = 0;
+
+  private static final IStatus MODIFYING_RESOURCES_ERROR_STATUS =
+      new Status(
+          IStatus.ERROR,
+          Saros.PLUGIN_ID,
+          ERROR_CODE,
+          Messages.ResourceChangeValidator_ModifyingResourcesErrorMessage,
+          null);
+
+  private static final IStatus DELETE_PROJECT_ERROR_STATUS =
+      new Status(
+          IStatus.ERROR,
+          Saros.PLUGIN_ID,
+          ERROR_CODE,
+          Messages.ResourceChangeValidator_DeleteProjectErrorMessage,
+          null);
+
+  private static final IStatus MOVE_OR_RENAME_PROJECT_ERROR_STATUS =
+      new Status(
+          IStatus.ERROR,
+          Saros.PLUGIN_ID,
+          ERROR_CODE,
+          Messages.ResourceChangeValidator_MoveOrRenameProjectErrorMessage,
+          null);
+
+  private static volatile ISarosSession session;
+
+  /* The provider is created on the fly when needed so we cannot benefit from the initialize method and so this has to be static. */
```

Comment:
Wouldn't it then be easier to hold a static reference to the session manager instead and just request the session object when needed? It would require a static block injecting the session manager, but would basically be the same logic and would avoid the usage of the support class (`ResourceChangeValidatorSupport`) and the volatile usage of the session (not using the static reference directly).

Reply:
This was the former approach. I changed it because now you have two options to disable this class. Either via the plugin.xml or via the context factory.

TBH I do not like this crappy plugin.xml file.

Reply:
I am not sure I understand what you mean. Could you elaborate? Do you mean that you can also disable this by simply not setting the session (e.g. by not starting the support class)? What advantages does this provide?